packages/expo-cli/src/accounts.ts

```diff
@@ -95,7 +110,171 @@ export async function login(options: CommandOptions): Promise<User> {
+  const cancelChoiceSentinel = -2;
+
+  const deviceChoices = smsNonPrimarySecondFactorDevices.map((device, idx) => ({
+    name: `Phone ending in ${device.sms_phone_number}`,
+    value: idx,
+  }));
+
+  if (hasOtherAuthenticatorSecondFactorDevice) {
+    deviceChoices.push({
+      name: 'Authenticator',
+      value: authenticatorChoiceSentinel,
+    });
+  }
+
+  deviceChoices.push({
+    name: 'Cancel',
+    value: cancelChoiceSentinel,
+  });
+
+  const question: Question = {
+    type: 'list',
+    name: 'choice',
+    message: 'Select a second factor device:',
+    choices: deviceChoices,
+    pageSize: Infinity,
+  };
+
+  const { choice } = await prompt(question);
+  if (choice === cancelChoiceSentinel) {
+    throw new Error('User cancelled login');
+  } else if (choice === authenticatorChoiceSentinel) {
+    return await _promptForOTPAsync('cancel');
+  }
+
+  const device = smsNonPrimarySecondFactorDevices[choice];
+
+  const apiAnonymous = ApiV2.clientForUser();
+  await apiAnonymous.postAsync('auth/send-sms-otp', {
+    username,
+    password,
+    secondFactorDeviceID: device.id,
+  });
+
+  return await _promptForOTPAsync('cancel');
+}
+
+/**
+ * Handle the special case error indicating that a second-factor is required for
+ * authentication.
+ *
+ * There are three cases we need to handle:
+ * 1. User's primary second-factor device was SMS, OTP was automatically sent by the server to that
+ *    device already. In this case we should just prompt for the SMS OTP (or backup code), which the
+ *    user should be receiving shortly. We should give the user a way to cancel and the prompt and move
+ *    to case 3 below.
+ * 2. User's primary second-factor device is authenticator. In this case we should prompt for authenticator
+ *    OTP (or backup code) and also give the user a way to cancel and move to case 3 below.
+ * 3. User doesn't have a primary device or doesn't have access to their primary device. In this case
+ *    we should show a picker of the SMS devices that they can have an OTP code sent to, and when
+ *    the user picks one we show a prompt for the sent OTP.
+ */
+async function _retryUsernamePasswordAuthWithOTPAsync(
+  username: string,
+  password: string,
+  metadata: {
+    secondFactorDevices?: SecondFactorDevice[];
+    smsAutomaticallySent?: boolean;
+  }
+): Promise<User> {
+  const {
+    secondFactorDevices,
+    smsAutomaticallySent,
+  }: {
+    secondFactorDevices?: SecondFactorDevice[];
+    smsAutomaticallySent?: boolean;
+  } = metadata as any;
+  invariant(
+    secondFactorDevices !== undefined && smsAutomaticallySent !== undefined,
+    'malformed OTP error metadata'
+  );
+
+  const primaryDevice = secondFactorDevices.find(device => device.is_primary);
+  let otp: string | null = null;
+
+  if (smsAutomaticallySent) {
+    invariant(
+      primaryDevice,
+      'OTP should only automatically be sent when there is a primary device'
+    );
+    console.log(
```

Comment:
I'm not familiar with this codebase but I would double check that `console.log` is the established way to write to the terminal instead of via some other logging API

Reply:
As far as I can tell via grepping it is the correct method.